src/tools/PICS-generator/PICSGenerator.py

```diff
@@ -0,0 +1,429 @@
+
+
+def cleanDirectory(pathToClean):
+    for entry in pathToClean.iterdir():
+        if entry.is_file():
+            pathlib.Path(entry).unlink()
+        elif entry.is_dir():
+            cleanDirectory(entry)
+            pathlib.Path(entry).rmdir()
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--cluster-data', required=True)
+parser.add_argument('--pics-template', required=True)
+parser.add_argument('--pics-output', required=True)
+args, unknown = parser.parse_known_args()
+
+basePath = os.path.dirname(__file__)
+clusterInfoInputPathStr = args.cluster_data
+
+xmlTemplatePathStr = args.pics_template
+if not xmlTemplatePathStr.endswith('/'):
+    xmlTemplatePathStr += '/'
+
+baseOutputPathStr = args.pics_output
+if not baseOutputPathStr.endswith('/'):
+    baseOutputPathStr += '/'
+
+serverTag = ".S"
+clientTag = ".C"
+featureTag = ".F"
+attributeTag = ".A"
+commandTag = ".C"
+acceptedCommandTag = ".Rsp"
+generatedCommandTag = ".Tx"
+
+# List of globale attributes (server)
+# Does not read ClusterRevision [0xFFFD] (not relevant), EventList [0xFFFA] (Provisional)
+featureMapAttributeId = "0xFFFC"
+attributeListAttributeId = "0xFFFB"
+acceptedCommandListAttributeId = "0xFFF9"
+generatedCommandListAttributeId = "0xFFF8"
+
+# Endpoint define
+rootNodeEndpointID = 0
+
+# Load cluster info
+inputJson = {}
+clusterInfoDict = {}
+
+print("Generating cluster data dict from JSON input")
+
+with open(clusterInfoInputPathStr, 'rb') as clusterInfoInputFile:
+    clusterInfoJson = json.load(clusterInfoInputFile)
+
+    for cluster in clusterInfoJson:
+        clusterData = clusterInfoJson[f"{cluster}"]["Data created by Script"]
+
+        try:
+            # Check if cluster id is a value hex value
+            clusterIdValid = int(clusterData["Id"].lower(), 16)
+
+            # Add cluster data to dict
+            clusterInfoDict[clusterData["Id"].lower()] = {
+                "Name": clusterData["Cluster Name"],
+                "PICS_Code": clusterData["PICS Code"],
+            }
+
+        except ValueError:
+            print(f"Ignore ClusterID: {clusterData['Id']} - {clusterData['Cluster Name']}")
+
+# Load PICS XML templates
+print("Capture list of PICS XML templates")
+xmlFileList = os.listdir(xmlTemplatePathStr)
+
+# Setup output path
+baseOutputPath = pathlib.Path(baseOutputPathStr)
+if not baseOutputPath.exists():
+    print("Create output folder")
+    baseOutputPath.mkdir()
+else:
+    print("Clean output folder")
+    cleanDirectory(baseOutputPath)
+
+
+class DeviceMappingTest(MatterBaseTest):
+    @async_test_body
+    async def test_device_mapping(self):
+
+        # Create console to print
+        global console
+        console = Console()
+
+        # Run device mapping function
+        await DeviceMapping(self.default_controller, self.dut_node_id, baseOutputPathStr)
+
+
+if __name__ == "__main__":
+    default_matter_test_main()
```

Comment:
I think you probably want all of the arg parsing handled in main. Does the argparse above fight with the argument parsing in the MatterBaseTest?

Reply:
No they should not collide, they are mainly additional stuff to get the paths to the "input" needed for the PICS generator to find what it needs.

Reply:
okie dokie.